packages/core/src/render3/util/defer.ts

```diff
@@ -0,0 +1,210 @@
+  const blocks: DeferBlockDetails[] = [];
+  getDeferBlocksInternal(lView, blocks);
+
+  for (const details of blocks) {
+    const native = getNativeByTNode(details.tNode, details.lView);
+    const lDetails = getLDeferBlockDetails(details.lView, details.tNode);
+
+    // The LView from `getLContext` might be the view the element is placed in.
+    // Filter out defer blocks that aren't inside the specified root node.
+    if (!node.contains(native as Node)) {
+      continue;
+    }
+
+    const tDetails = details.tDetails;
+    const renderedView = getRendererView(details);
+    const data: DeferBlockData = {
+      state: stringifyState(lDetails[DEFER_BLOCK_STATE]),
+      hydrationState: inferHydrationState(tDetails, lDetails, registry),
+      hasErrorBlock: tDetails.errorTmplIndex !== null,
+      loadingBlock: {
+        exists: tDetails.loadingTmplIndex !== null,
+        minimumTime: tDetails.loadingBlockConfig?.[0] ?? null,
+        afterTime: tDetails.loadingBlockConfig?.[1] ?? null,
+      },
+      placeholderBlock: {
+        exists: tDetails.placeholderTmplIndex !== null,
+        minimumTime: tDetails.placeholderBlockConfig?.[0] ?? null,
+      },
+      triggers: tDetails.debug?.triggers ? Array.from(tDetails.debug.triggers) : [],
+      rootNodes: renderedView === null ? [] : getRootNodes(renderedView),
+    };
+
+    results.push(data);
+
+    // `getDeferBlocks` does not resolve nested defer blocks so we have to recurse manually.
+    if (renderedView !== null) {
+      findDeferBlocks(node, renderedView, results);
+    }
+  }
+}
+
+/**
+ * Turns the `DeferBlockState` into a string which is more readable than the enum form.
+ *
+ * @param lDetails Information about the
+ * @returns
+ */
+function stringifyState(state: DeferBlockState | DeferBlockInternalState): DeferBlockData['state'] {
+  switch (state) {
+    case DeferBlockState.Complete:
+      return 'complete';
+    case DeferBlockState.Loading:
+      return 'loading';
+    case DeferBlockState.Placeholder:
+      return 'placeholder';
+    case DeferBlockState.Error:
+      return 'error';
+    case DeferBlockInternalState.Initial:
+      return 'initial';
+  }
+
+  throw new Error(`Unrecognized state ${state}`);
+}
+
+/**
+ * Infers the hydration state of a specific defer block.
+ * @param tDetails Static defer block information.
+ * @param lDetails Instance defer block information.
+ * @param registry Registry coordinating the hydration of defer blocks.
+ */
+function inferHydrationState(
+  tDetails: TDeferBlockDetails,
+  lDetails: LDeferBlockDetails,
+  registry: DehydratedBlockRegistry | null,
+): DeferBlockData['hydrationState'] {
+  if (
+    registry === null ||
+    lDetails[SSR_UNIQUE_ID] === null ||
+    tDetails.hydrateTriggers === null ||
+    tDetails.hydrateTriggers.has(DeferBlockTrigger.Never)
+  ) {
+    return 'not-hydratable';
+  }
+  return registry.has(lDetails[SSR_UNIQUE_ID]) ? 'dehydrated' : 'hydrated';
```

Comment:
I'm a bit unsure about the states: a block might not use incremental hydration (e.g. do not have `hydrate` triggers set up), thus the placeholder would be SSR'ed and later hydrated on the client. I think we can make it more explicit that this is an incremental hydration state by naming a corresponding property `incrementalHydrationState` (vs `hydrationState`) and renaming one of the return values:

```suggestion
    return 'not-configured';
  }
  return registry.has(lDetails[SSR_UNIQUE_ID]) ? 'dehydrated' : 'hydrated';
```